### DISCUSSION TITLE
Create conditional Wrapper/Passthrough Element

### DISCUSSION THREAD

**Discussion by phonzammi** [vikejs/vike#1516] [Discussion] [answered]
Hi @brillout,
I have question and asking for suggestion about conditional wrapper/passthrough element.

For starters.
I have scaffold vike-react with [https://batijs.github.io/], then i try to integrate it with tanstack-query. I did prefetching movies data in pages/star-wars/index/+onBeforeRender, then dehydrate and return pageProps: {dehydratedState}

I could wrap HydrationBoundary in the Page that has dehydratedState, but i don't want to repeat it.

I can't just wrap children in Layout Component with HydrationBoundary, because not every Page have pageProps.dehydratedState

If i am not mistaken, i've seen the PassThrough element from vike/example but i can't find it anymore.
So i try to create it my self, here is the code

```
export default function LayoutDefault({ children }) {
  const pageContext = usePageContext()
  const { pageProps } = pageContext
  const PassThrough = pageProps && pageProps.dehydratedState ?
    () => <HydrationBoundary state={pageProps.dehydratedState}>{children}</HydrationBoundary>
    : () => <>{children}</>
  return (
    <div className="flex max-w-5xl m-auto">
      <Sidebar>
        <Logo />
        <Link href="/">Welcome</Link>
        <Link href="/star-wars">Data Fetching</Link>
      </Sidebar>
      <Content>
        <RQProviders>
          <PassThrough />
        </RQProviders>
      </Content>
    </div>
  );
}
```

**Comment by brillout** [comment, marked as answer]
The thing is that @AurelienLourot and myself aren't sure whether the slightly improved DX is worth an extra `config.Wrapper` making the overall interface surface slightly less minimal.

For now, adding your own wrapper to your layouts is the way to go. (Your code seems fine to me, although I'd rename `PassThrough` to `Wrapper`.)

Would you be up for creating a vike-react + TranStack Query example? We could then polish it together. (Maybe we could even create a `vike-react-query` 🤔)

**Reply by phonzammi** [reply]
> The thing is that @AurelienLourot and myself aren't sure whether the slightly improved DX is worth an extra `config.Wrapper` making the overall interface surface slightly less minimal.

I don't think i need an extra config.Wrapper right now, but maybe can we use that extra config.Wrapper if we want to integrate vike-react with Graphql? 🤔

> For now, adding your own wrapper to your layouts is the way to go. (Your code seems fine to me, although I'd rename `PassThrough` to `Wrapper`.)

Are you sure ? Thanks 🥳. But I still curious, maybe you have cleaner/better code 🤩

> Would you be up for creating a vike-react + TranStack Query example? We could then polish it together. (Maybe we could even create a vike-react-query 🤔)

Sure, I am doing it for almost a month & still trying, I prefer Vike v1 design (with or without builtin renderer) & latest Tanstack Query.

**Reply by phonzammi** [reply]
> Would you be up for creating a vike-react + TranStack Query example? We could then polish it together. (Maybe we could even create a vike-react-query 🤔)

I also have another question about prefetching. I want to know your opinion where I am wrong, and please tell me the right.
I think we don't need to prefetch everything or maybe anything inside the renderer (`onRenderClient` or `onRenderHtml`).

As usual (from the docs) we do fetching data inside `onBeforeRender`.
 I think that is also the best place to do prefetching data inside `onBeforeRender` of a page that needs prefetching.
So we don't fill our html with bunch of data from another page, right?

The main purpose is to do prefetching on the server (inside `onBeforeRender`), than we can use useQuery hook inside +Page and data will automatically be there, let the client take care of the rest (refetching, invalidation, etc).

```
/layout/LayoutDefault.tsx ---> QueryClientProvider > HydrationBoundary state={dehydratedState} (if returned from `onBeforeRender`)
/pages/star-wars/index/+Page.tsx ---> useQuery > isLoading always false, can do refetching without reload/rerender
/pages/star-wars/index/+onBeforeRender.ts ---> queryClient.prefetch() ---> dehydrate(queryClient) return to pageProps : dehydratedState
```

Sorry if its too long, so what do you think @brillout, and you too @AurelienLourot 🧐 🤔

**Reply by lourot** [reply]
> I think we don't need to prefetch everything or maybe anything inside the renderer (`onRenderClient` or `onRenderHtml`).
> 
> As usual (from the docs) we do fetching data inside `onBeforeRender`. I think that is also the best place to do prefetching data inside `onBeforeRender` of a page that needs prefetching.

I have just wrapped my head around https://tanstack.com/query/latest/docs/react/guides/ssr#using-the-hydration-apis and yes I agree: prefetching returning dehydrated client should be done in `onBeforeRender`. At least that's what I would intuitively do.

> `/layout/LayoutDefault.tsx ---> QueryClientProvider > HydrationBoundary state={dehydratedState}`

Yes this sounds right to me, as far as I can tell without trying things out myself. Thanks for looking into this!

**Reply by phonzammi** [reply]
Wow @AurelienLourot  
Thanks for that, 🥳

Actually it's been too long I want to discuss about this with you guys, but I was embarrassed due to lack of knowledge, and my english is bad, btw I am from Indonesia.

I hope i am on the right track, i have been reading tanstack query docs since v4 to v5 about ssr & hydration, also searching & watching everything related to react, tanstack query & ssr. I don't know next js or remix, I don't want to learn them both, not yet. I have been thinking and trying how to integrate it with Vike with the best practice.

Sorry to disturb your time @brillout & @AurelienLourot.
And please help me figure this out
